### ISSUE TITLE
Bug: user timezones don't get updated with daylight savings time changes

### ISSUE THREAD

**Issue by isabensusan** [open-sauced/app#2312]
### Describe the bug

Jesse from Xebia brought this up:
There was a time shift in the Netherlands for winter time (same logic as daylight savings in the US), but the local time displayed in user's profiles didn't get updated.

> The TZ on myprofile still shows UTC2, but since it's CET-Winter-time (Amsterdam), it should show UTC1 at the moment.

Jesse's example -- his profile shows that the time for him in Amsterdam is 7:37pm, but if you look at the actual time in Amsterdam at the moment it's 6:37pm because of the time shift (these two images were taken two minutes apart):

<img width="424" alt="image" src="https://github.com/open-sauced/app/assets/27322879/4d372dbc-5dd4-4012-93a3-3c790cb30328">

<img width="408" alt="image" src="https://github.com/open-sauced/app/assets/27322879/accc9e65-0d87-42ff-aef8-f262ac60cff8">

### Steps to reproduce

1. Go to Jesse's profile: https://app.opensauced.pizza/user/jessehouwing
2. See the local time info
3. Google "time in Amsterdam"
4. See time difference

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [X] I agree to follow this project's Contribution Docs

**Comment by nickytonline** [comment]
We have the list of timezones hard-coded which do not take changes into account.

https://github.com/open-sauced/app/blob/cef3efc502ff664692a7ca06e03c1797fae41b56/lib/utils/timezones.ts#L3-L1181

I'd suggest (and I think we discussed this at some point @jpmcb) that hitting an API would be better and we could cache the list for an indeterminate amount of time.

timezonedb.com or maybe the Google timezone API. Open to others recommendations.

**Comment by bdougie** [comment]
> timezonedb.com or maybe the Google timezone API. Open to others recommendations.

Seems like the Google timezone has what we needed, but we should consider timezonedb at a later date if we need that flexibility. I don't have experience with paying for an API like this, and it is not clear if commercials use cost more.

**Comment by nickytonline** [comment]
Yeah Google is probably a safe bet, but I’ll let others chime in.

**Comment by jpmcb** [comment]
> I'd suggest (and I think we discussed this at some point @jpmcb) that hitting an API would be better and we could cache the list for an indeterminate amount of time.

Yes, that's part of the puzzle. But this would encompass a much bigger change to the way our data is structure and how we encapsulate user's timezones: right now, we store timezones in the API as raw text fields (like "Mountain Standard Time" in my user's row or "Central Standard Time" in Brandon's row in the database). I _believe_ these are propagating through from this raw list of timezones in the frontend you mentioned.

Those should probably be UTC offset timezones (like UTC−07:00 for mountain time) since they are totally agnostic to different jurisdiction's interpretation of timezones.

Then, with a UTC offset, we could hit any number of different time sync services ([like amazon has a public NTP server we could hit](https://aws.amazon.com/about-aws/whats-new/2022/11/amazon-time-sync-internet-public-ntp-service/), although probably overkill for our purposes) to get the right localized timezone.

So, TLDR: there's more work to get this correct from the data perspective than just changing these types.

**Comment by davidlj95** [comment]
Hi! I was surprised to see `(UTC+01:00)` when selecting Madrid's time zone in the app and came across this issue. I've dealt with time in some apps in the pass and is definitely a [PIA](https://en.wiktionary.org/wiki/pain_in_the_ass) 🥲 Some things I can say from my experience after reading the issue and the hardcoded timezones code:

## Issues
### Conceptual: UTC offset != time zone
As [Moment.js explains on their site](https://momentjs.com/guides/#/lib-concepts/timezone-offset/)

> A UTC offset is a value that represents how far a particular date and time is from UTC. It is expressed in the format HH:mm most of the time.
>
> A time zone is a geographical region where all people observe a legally mandated standard time.
>
> A time zone usually has more than one offset from UTC due to daylight saving time.
> [_Source code in case website link breaks_](https://github.com/moment/momentjs.com/blob/5702dc332ab9676e9f3b3e3cca3c6c7cb21ec6a8/guides/moment/00-lib-concepts/03-timezone-offset.md)

So when saying for instance:

> Then, with a UTC offset, we could hit any number of different time sync services [...] to get the right localized timezone.

It's not possible! Because many people in different parts of the world may be in the same UTC offset (some time ahead/behind UTC). That's [misconception 2 about time zones][falsehoods-timezones].

[falsehoods-timezones]: https://www.zainrizvi.io/blog/falsehoods-programmers-believe-about-time-zones/

A weird case is for instance is Newfoundland, a city in Canada. Despite [most of Canada uses UTC offsets from -07:00 to -04:00](https://upload.wikimedia.org/wikipedia/commons/8/88/World_Time_Zones_Map.png) [the offset they use is -03:30](https://en.wikipedia.org/wiki/Newfoundland_Time_Zone) 🙃 There are also [funny time zones around](https://citymonitor.ai/government/the-worlds-weirdest-time-zones)

Mindblown? 🤯  Well, the above paragraph and linked map are not 100% correct either though 🤓 Because it's not taking into account daylight saving time (DST). When it's active, Canadians use UTC offsets from -07:00 to -03:00 ([-07:00 for some given they do not have DST](https://www.timeanddate.com/news/time/usa-canada-end-dst-2023.html)). Except Newfoundlanders, who use a -02:30 UTC offset for their time.

So to summarize...

- **UTC offset**: a fixed amount of time behind / ahead from UTC. 
- **Time zone**: a geographical area where people have agreed on certain rules about time. 
 
Time zone data allows to calculate UTC offsets from there. Those UTC offsets may vary due to different reasons, like DST. And that therefore includes the time instant to localize. A date in winter won't have DST whilst a date in summer will. Can be others too. Like for instance going from using DST to not using DST at a certain point in time (or viceversa 😛 ).

### Specific: hardcoded time zones & UTC offsets

Right now the hardcoded time zones file contains a hardcoded UTC offset and a text containing it. This is my timezone for instance:

https://github.com/open-sauced/app/blob/eb4d77fab6dc797fdba57ed6d33e9d5f292a710b/lib/utils/timezones.ts#L538-L545

Given we're right now in daylight saving time (DST), the UTC offset is actually `(UTC+02:00)` and not `(UTC+01:00)`. The `offset` field is correct though. In this case it's due to DST. But as mentioned in the issue, an external source should be used for that in order to keep updated with the available time zones and their rules to get UTC offset for a given time moment.

## Solutions
First of all, using time zones and not UTC offsets when localizing a time instant (ie: a [Unix time](https://en.wikipedia.org/wiki/Unix_time) stamp or a date in UTC). As explained above, a UTC offset is a different concept from a time zone. Having stored time zone names in the API is a great choice then.

Where to fetch time zones data though? Given all of that mess, querying an API like proposed  [Google's Time Zone API](https://developers.google.com/maps/documentation/timezone/overview) one makes sense to have all those rules up to date. However, those rules do not change so often. So using a time zones database can save 1 precious API query round trip time. Most used is [IANA's Time Zone Database](https://www.iana.org/time-zones) AFAIK. Kind of an industry standard as used in many many softwares around directly or indirectly. Updating the library providing this database from time to time would be more than enough to keep those time rules updated.

Talking about libraries, I saw `date-fns` is used. Seems that [`date-fns-tz`](https://www.npmjs.com/package/date-fns-tz) brings time zones support to `date-fns`. Seems it uses [browsers' `Intl` API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) so will indeed reduce your bundle size given you won't need to serve that database as it's using the user's browsers one\*! There are polyfills also if support for old browsers is needed.

> Which probably will use OS' one. Which probably will be IANA's one. Unless if using Windows which they use their own one AFAIK.

I'll give it a try and raise a PR if have some time 😃 

> Apologies for the bible being written here, I get confused about this topic often, so writing about it helps training myself about it (and about writing too 😛 ) 

## Cool resources about managing time
Some links that have been useful about the topic:

- [The Problem with Time & Timezones - Computerphile ~Tom Scott getting mad due to time & time zones~](https://www.youtube.com/watch?v=-5wpm-gesOY)
- [Falsehoods programmers believe about time zones][falsehoods-timezones]
- [Falsehoods programmers believe about time](https://infiniteundo.com/post/25326999628/falsehoods-programmers-believe-about-time)
